numba_dpex/caching.py

```diff
@@ -0,0 +1,129 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+
+from numba.core.caching import IndexDataCacheFile, _Cache, _CacheImpl
+from numba.core.serialize import dumps
+
+from numba_dpex.core import compiler
+
+
+class DpexCacheImpl(_CacheImpl):
+    def reduce(self, cres):
+        """
+        Returns a serialized CompileResult
+        """
+        return cres._reduce()
+
+    def rebuild(self, target_context, payload):
+        """
+        Returns the unserialized CompileResult
+        """
+        return compiler.CompileResult._rebuild(target_context, *payload)
+
+    def check_cachable(self, cres):
+        # For the time being, assuming all numba-dpex Kernels are always cachable.
+        return True
+
+
+class DpexCache(_Cache):
+    """
+    Implements a cache that saves and loads CUDA kernels and compile results.
+    """
+
+    _impl_class = DpexCacheImpl
+
+    def __init__(self, py_func):
+        self._name = repr(py_func)
+        self._py_func = py_func
+        self._impl = self._impl_class(py_func)
+        self._cache_path = self._impl.locator.get_cache_path()
+        # This may be a bit strict but avoids us maintaining a magic number
+        source_stamp = self._impl.locator.get_source_stamp()
+        filename_base = self._impl.filename_base
+        self._cache_file = IndexDataCacheFile(
+            cache_path=self._cache_path,
+            filename_base=filename_base,
+            source_stamp=source_stamp,
+        )
+        self.enable()
+
+    @property
+    def cache_path(self):
+        return self._cache_path
+
+    def enable(self):
+        self._enabled = True
+
+    def disable(self):
+        self._enabled = False
+
+    def flush(self):
+        self._cache_file.flush()
+
+    def load_overload(self, sig, target_context):
+        if not self._enabled:
+            return
+        # key = self._index_key(sig, target_context.codegen())
+        key = self._index_key(sig)
+        data = self._cache_file.load(key)
+        if data is not None:
+            data = self._impl.rebuild(target_context, data)
+        print("-----> caching.load_overload()")
+        return data
+
+    def save_overload(self, sig, data):
+        if not self._enabled:
+            return
+        if not self._impl.check_cachable(data):
+            return
+        self._impl.locator.ensure_cache_path()
+        # key = self._index_key(sig, data.codegen)
+        key = self._index_key(sig)
+        # data = self._impl.reduce(data)
```

Comment:
@diptorupd This `reduce` doesn't work on the compiled kernel either. I am getting this:
```bash
python driver.py
kernel: 2
-----> Dispatcher.__init__()
/localdisk/work/akmkhale/numba-dpex/driver.py:28: DeprecationWarning: The [] (__getitem__) method to set global and local ranges for launching a kernel is deprecated. Use the execute function instead.
  data_parallel_sum[(100,)](a, b, c)
/localdisk/work/akmkhale/numba-dpex/numba_dpex/core/kernel_interface/dispatcher.py:334: UserWarning: Use of __getitem__ to set the global_range attribute is deprecated. Use the keyword argument "global_range" of __call__ method to set the attribute.
  warn(
/localdisk/work/akmkhale/numba-dpex/numba_dpex/core/kernel_interface/dispatcher.py:359: UserWarning: Kernel to be submitted without a local range letting the SYCL runtime select a local range. The behavior can lead to suboptimal performance in certain cases. Consider setting the local range value for the kernel execution.
The local_range keyword may be made a required argument in the future.
  warn(
-----> dispatcher.kernel_name: data_parallel_sum
-----> spriv_kernel._compile()
-----> caching.load_overload()
-----> spirv_kernel._compile().cres == None
Traceback (most recent call last):
  File "/localdisk/work/akmkhale/numba-dpex/driver.py", line 36, in <module>
    main()
  File "/localdisk/work/akmkhale/numba-dpex/driver.py", line 28, in main
    data_parallel_sum[(100,)](a, b, c)
  File "/localdisk/work/akmkhale/numba-dpex/numba_dpex/core/kernel_interface/dispatcher.py", line 427, in __call__
    kernel.compile(
  File "/localdisk/work/akmkhale/numba-dpex/numba_dpex/core/kernel_interface/spirv_kernel.py", line 167, in compile
    cres = self._compile(
  File "/nfs/site/home/akmkhale/.conda/envs/numba-dpex/lib/python3.9/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/localdisk/work/akmkhale/numba-dpex/numba_dpex/core/kernel_interface/spirv_kernel.py", line 120, in _compile
    self._cache.save_overload(cres.signature, cres)
  File "/localdisk/work/akmkhale/numba-dpex/numba_dpex/caching.py", line 95, in save_overload
    data = self._impl.reduce(data)
  File "/localdisk/work/akmkhale/numba-dpex/numba_dpex/caching.py", line 28, in reduce
    return cres._reduce()
  File "/nfs/site/home/akmkhale/.conda/envs/numba-dpex/lib/python3.9/site-packages/numba/core/compiler.py", line 183, in _reduce
    libdata = self.library.serialize_using_object_code()
  File "/nfs/site/home/akmkhale/.conda/envs/numba-dpex/lib/python3.9/site-packages/numba/core/codegen.py", line 922, in serialize_using_object_code
    data = (self._get_compiled_object(),
  File "/nfs/site/home/akmkhale/.conda/envs/numba-dpex/lib/python3.9/site-packages/numba/core/codegen.py", line 630, in _get_compiled_object
    raise RuntimeError("no compiled object yet for %s" % (self,))
RuntimeError: no compiled object yet for <Library 'data_parallel_sum' at 0x7f1f1c4fbcd0>
```